identity/src/test/java/com/google/enterprise/cloudsearch/sdk/identity/externalgroups/ExternalGroupsRepositoryTest.java

```diff
@@ -0,0 +1,277 @@
+        + "  \"members\":[ {\"id\":\"customer-principal-group@example.com\"} ]},"
+        + " {\"name\":\"BUILTIN\\\\Administrators\","
+        + "  \"members\":[ {\"id\":\"admin-group\","
+        + "                 \"referenceIdentitySourceName\":\"idSourceForFileShare\"} ]}"
+        + "]}";
+    List<IdentityGroup> expected = Arrays.asList(
+        makeIdentityGroup(identitySourceId, "Everyone",
+            makeEntityKey("customer-principal-group@example.com")),
+        makeIdentityGroup(identitySourceId, "BUILTIN\\Administrators",
+            makeEntityKey("admin-group", "identitysources/9876543211")));
+
+    Properties config = new Properties();
+    config.put("api.identitySourceId", identitySourceId);
+    config.put("api.referenceIdentitySources", "idSourceForFileShare, idSourceForHogwarts");
+    config.put("api.referenceIdentitySource.idSourceForFileShare.id", "9876543211");
+    config.put("api.referenceIdentitySource.idSourceForHogwarts.id", "77777777");
+    config.put("externalgroups.filename", temporaryFolder.newFile("unused").toString());
+    setupConfig.initConfig(config);
+    RepositoryContext repositoryContext = RepositoryContext.fromConfiguration();
+    ExternalGroupsRepository groupsRepository =
+        new ExternalGroupsRepository(() -> fromString(testGroups));
+    groupsRepository.init(repositoryContext);
+
+    List<IdentityGroup> groups = getAll(groupsRepository.listGroups(null));
+    assertEquals(2, groups.size());
+    assertEquals(expected, groups);
+  }
+
+  @Test
+  public void listGroups_referenceIdentitySource_missingConfig() throws Exception {
+    String identitySourceId = "test-identitySourceId";
+    final String testGroups =
+        "{\"externalGroups\":["
+        + " {\"name\":\"BUILTIN\\\\Administrators\","
+        + "  \"members\":[ {\"id\":\"admin-group\","
+        + "                 \"referenceIdentitySourceName\":\"no-such-source\"} ]}"
+        + "]}";
+
+    Properties config = new Properties();
+    config.put("api.identitySourceId", identitySourceId);
+    config.put("externalgroups.filename", temporaryFolder.newFile("unused").toString());
+    setupConfig.initConfig(config);
+    RepositoryContext repositoryContext = RepositoryContext.fromConfiguration();
+    ExternalGroupsRepository groupsRepository =
+        new ExternalGroupsRepository(() -> fromString(testGroups));
+    groupsRepository.init(repositoryContext);
+
+    thrown.expect(InvalidConfigurationException.class);
+    groupsRepository.listGroups(null);
+  }
+
+  @Test
+  public void listUsers_throwsException() throws Exception {
+    ExternalGroupsRepository groupsRepository = new ExternalGroupsRepository();
+    thrown.expect(UnsupportedOperationException.class);
+    groupsRepository.listUsers(null);
+  }
+
+  @Test
+  public void close_doesNothing() throws Exception {
+    ExternalGroupsRepository groupsRepository = new ExternalGroupsRepository();
+    groupsRepository.close();
```

Comment:
Does it make any sense to use groupsRepository after this, to actually demonstrate the no-op? Or is that an overspecification?